ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -0,0 +1,517 @@
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+            >::new();
+            // first lagrange table
+            let first_lagrange_table = LagrangeTable::<
+                Fp61BitPrime,
+                { LargeProofGenerator::RECURSION_FACTOR },
+                { LargeProofGenerator::LAGRANGE_LENGTH },
+            >::from(first_denominator);
+
+            LargeProofGenerator::compute_next_proof(
+                ctx,
+                &mut record_counter,
+                &first_lagrange_table,
+                Self::polynomials_from_blocks(uv_tuple_block),
+            )
+        };
+
+        // storage for other proofs
+        let mut prover_left_proofs =
+            Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+        let mut proofs_from_left = Vec::<[Fp61BitPrime; SmallProofGenerator::PROOF_LENGTH]>::new();
+
+        // precomputation for other proofs
+        // denominator
+        let denominator = CanonicalLagrangeDenominator::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+        >::new();
+        // lagrange table
+        let lagrange_table = LagrangeTable::<
+            Fp61BitPrime,
+            { SmallProofGenerator::RECURSION_FACTOR },
+            { SmallProofGenerator::LAGRANGE_LENGTH },
+        >::from(denominator);
+
+        // recursively generate proofs
+        // until there are less than `RECURSION_FACTOR * (RECURSION_FACTOR-1)` many (u,v) field elements
+        // and then do one more iteration
+        // This ensures that after the last iteration,
+        // there are at most RECURSION_FACTOR-1 many real (u,v) field elements
+        // Therefore at least the last field element is a `F:ZERO` filler that can be replaced
+        // Therefore we have space to include the masks during the final proof
+        //
+        // we remark that uv_value.len() will always be at least 1
+        // further, polynomial uv_values.uv[0].0 and .1 will always have
+        // RECURSION_FACTOR many points
+        // since they are arrays of fixed length, i.e. RECURSION_FACTOR,
+        // (which are filled with `F::ZERO` when there are not enough elements).
+        let r_minus_one = SmallProofGenerator::RECURSION_FACTOR - 1;
+        uv_values = loop {
+            let stop = uv_values.len() <= r_minus_one;
+
+            // generate next proof
+            let (uv_values_new, proof_from_left, prover_left_proof) =
+                SmallProofGenerator::compute_next_proof(
+                    ctx,
+                    &mut record_counter,
+                    &lagrange_table,
+                    uv_values.iter(),
+                );
+            // collect proof
+            proofs_from_left.push(proof_from_left);
+            prover_left_proofs.push(prover_left_proof);
+
+            uv_values = uv_values_new;
+
+            if stop {
+                break uv_values;
+            }
+        };
```

Comment:
This logic is too subtle and complex. This will make it brittle and hard to maintain, and difficult for readers to understand. We need to break this apart and find a new approach to solve this problem that is implemented in simpler code.